### PR TITLE
Improvements to ActiveRecord::Rollback docs

### PR DESCRIPTION
**Pull request by jjb**
The main problem I wanted to solve is that the "Call tech support" message in the example is
confusing in this scenario, because it's addressing the user, but the user will never see it.
I'm not sure if I'm not understanding something, or maybe this was copy/pasted from
another example years ago.

Here I remove that and also expand the docs, hopefully illustrating the concept better.

Any and all feedback on copy, phrasing, formatting, examples, is welcome!

### CODE REVIEW

**Comment by p8** [comment]
Hey @jjb, thanks for the PR!

I can see how the  "Call tech support" line is confusing.
The example is getting a lot complexer though now.

What do you think about adding the `If this fails, user gets generic 500 behavior` comment above the `book.save!`
and removing the "Call tech support" string like you did?

**Comment by jjb** [comment]
done - if that looks good i can squish the two commits

> The example is getting a lot complexer though now

by this do you mean it's difficult to understand, or it's not emblematic of intended use case, or something else?

I think the docs currently in main are not emblematic of a typical use case, if you agree i wonder if we can improve the docs without making them too complex.

**Comment by p8** [comment]
@jjb adding the Author and Stats models makes things more complex.
Also, while the different responses ("rendering" and "redirecting") might be more realistic, they add more complexity that maybe isn't relevant to the example.

**Review by p8 — APPROVED** [review]
This looks good to me! 🚢 
Please squash the commits.

**Comment by jjb** [comment]
@p8 squashed

File: activerecord/lib/active_record/errors.rb [hunk 1]
```diff
@@ -273,12 +273,13 @@ class StrictLoadingViolationError < ActiveRecordError
   #     def create
   #       Book.transaction do
   #         book = Book.new(params[:book])
+  #         # If this fails, user gets generic 500 behavior
```
Comment (p8) [inline activerecord/lib/active_record/errors.rb:276]:
```suggestion
  #         # If this fails, the user gets generic 500 behavior.
```

What do you think about making the 500 more explicit?
For beginners who don't know 500 is a HTTP code.
```suggestion
  #         # If this fails, the user gets the 500 server error behavior.
```
or maybe:
```suggestion
  #         # If this fails, the user sees a "500 Server Error" message.
```

**Review by p8 — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/errors.rb (hunk 1 above)
Comment (jjb) [reply activerecord/lib/active_record/errors.rb]: perhaps problem with this is, user won't see that copy, because app is likely configured to show something nicer

File: activerecord/lib/active_record/errors.rb [hunk 2]
```diff
@@ -273,12 +273,13 @@ class StrictLoadingViolationError < ActiveRecordError
   #     def create
   #       Book.transaction do
   #         book = Book.new(params[:book])
+  #         # If this fails, the user gets generic 500 behavior.
```
[the remark this answers is not in the document]
Comment (p8) [reply activerecord/lib/active_record/errors.rb:276]: In this case we can remove `book = Book.new(params[:book])`

[the remark this answers is not in the document]
Comment (jjb) [reply activerecord/lib/active_record/errors.rb]: cool- i'm not totally clear what you want to keep from 08cebde though

I just put the original code back in, maybe make suggestions there?

[the remark this answers is not in the document]
Comment (p8) [reply activerecord/lib/active_record/errors.rb]: Ah sorry I wasn't clear 🙂 
The TLDR is I'm suggesting to maybe change it to my code suggestion.

Your example had multiple models (`Stats` and `Author`). This is more realistic example for using a transaction.
where you want to rollback multiple operations. This can make the example clearer as we can comment "we rollback this transaction and pass on the exception", which differs from how `ActiveRecord::Rollback` works.
I don't think we need to make the example more complex with different types of rendering and redirecting.

In the suggested code above, I added your `Stats.create!(type: 'book_purchase')` line (I removed the params because I don't think it adds much to the example).
Adding an extra `create` to the transaction makes it clearer that the `Book.create` also gets rolled back.
To keep it simple I don't think with need to add `Author` as well.

Removing the `!` from `Book.create(params[:book])` makes sure that line doesn't raise.
If that is the first database operation nothing gets rolled back. Letting it raise could be confusing (does it get rolled back or is it never saved?).

I moved the `Stats.create!` below the `Book.create` to make it clear `Book.transaction` can also be rolled back by other classes than `Book`

**Comment by jjb** [comment]
@p8 okay, done

would you consider replacing the today_is_friday? joke example for when the rollback is used? I feel it's not a good way to explain a possible real world use case and the jokiness creates a cognitive speed bump while trying to understand.

**Comment by p8** [comment]
I’m okay with that. I like some light hearted reading, but it shouldn’t be at the cost of understandability. 

Maybe we should show the example in a model instead of a controller. An Order that requires the Book to be in the inventory? This allows us to skip all the rendering logic and use a more realistic example?

**Comment by jjb** [comment]
in that scenario, which operations should raise and which operations should continue on?

**Comment by jjb** [comment]
Hi @p8 - if you have time and interest to answer my question I can maybe put something together, thanks!

**Comment by p8** [comment]
Hi @jjb sorry for the late reply 😞 

Maybe we can render a different flash depending on wether the book was saved?
```ruby
  #   class BooksController < ActionController::Base
  #     def create
  #       book = Book.new(params[:book])
  #       Book.transaction do
  #         book.save!
  #         # If this fails, we rollback this transaction and pass on the exception.
  #         # The user gets generic 500 behavior.
  #         Stats.create!(type: 'book_purchase')
  #         if today_is_friday?
  #           # The system must fail on Friday so that our support department
  #           # won't be out of job. We silently rollback this transaction
  #           # without telling the user.
  #           raise ActiveRecord::Rollback
  #         end
  #       end
  #       # ActiveRecord::Rollback is the only exception that won't be passed on
  #       # by ActiveRecord::Base.transaction, so this will still be reached
  #       # even on Friday.
  #       if book.persisted?
  #         redirect_to root_url, notice: "Book purchased"
  #       else
  #         redirect_to root_url, alert: "Call tech support!"
  #       end
  #     end
  ```

**Comment by jjb** [comment]
thanks for looping back - i won't be able to work on this unfortunately - maybe someone else can pick it up and work with what i've suggested
